fdbrpc/JsonWebKeySet.cpp

```diff
@@ -0,0 +1,828 @@
+		alg = keyObj.algorithm();
+	} else {
+		auto const& keyObj = key.getPrivate();
+		pKey = key.getPrivate().nativeHandle();
+		alg = keyObj.algorithm();
+	}
+	if (!pKey) {
+		JWK_WRITE_ERROR("PKey object to encode is null");
+		return false;
+	}
+	if (alg == PKeyAlgorithm::EC) {
+		return encodeEcKey(writer, keyName, pKey, isPublic);
+	} else if (alg == PKeyAlgorithm::RSA) {
+		return encodeRsaKey(writer, keyName, pKey, isPublic);
+	} else {
+		JWK_WRITE_ERROR("Attempted to encode PKey with unsupported algorithm");
+		return false;
+	}
+	return true;
+}
+
+} // anonymous namespace
+
+Optional<JsonWebKeySet> JsonWebKeySet::parse(StringRef jwksString, VectorRef<StringRef> allowedUses) {
+	auto d = rapidjson::Document();
+	d.Parse(reinterpret_cast<const char*>(jwksString.begin()), jwksString.size());
+	if (d.HasParseError()) {
+		JWKS_PARSE_ERROR("ParseError")
+		    .detail("Message", GetParseError_En(d.GetParseError()))
+		    .detail("Offset", d.GetErrorOffset());
+		return {};
+	}
+	auto keysItr = d.FindMember("keys");
+	if (!d.IsObject() || keysItr == d.MemberEnd() || !keysItr->value.IsArray()) {
+		JWKS_PARSE_ERROR("JWKS must be an object and have 'keys' array member");
+		return {};
+	}
+	auto const& keys = keysItr->value;
+	auto ret = JsonWebKeySet{};
+	for (auto keyIndex = 0; keyIndex < keys.Size(); keyIndex++) {
+		if (!keys[keyIndex].IsObject()) {
+			JWKS_PARSE_ERROR("element of 'keys' array must be an object");
+			return {};
+		}
+		auto const& key = keys[keyIndex];
+		DECLARE_JWK_REQUIRED_STRING_MEMBER(key, kty);
+		DECLARE_JWK_REQUIRED_STRING_MEMBER(key, kid);
+		DECLARE_JWK_OPTIONAL_STRING_MEMBER(key, use);
+		if (use.present() && !allowedUses.empty()) {
+			auto allowed = false;
+			for (auto allowedUse : allowedUses) {
+				if (allowedUse == use.get()) {
+					allowed = true;
+					break;
+				}
+			}
+			if (!allowed) {
+				JWK_PARSE_ERROR("Illegal optional 'use' member found").detail("Use", use.get().toString());
+				return {};
+			}
+		}
+		auto parsedKey = parseKey(key, kty, keyIndex);
+		if (!parsedKey.present())
+			return {};
+		auto [iter, inserted] = ret.keys.insert({ Standalone<StringRef>(kid), parsedKey.get() });
+		if (!inserted) {
+			JWK_PARSE_ERROR("Duplicate key name").detail("KeyName", kid.toString());
+			return {};
+		}
+	}
+	return ret;
+}
+
+Optional<StringRef> JsonWebKeySet::toStringRef(Arena& arena) {
+	using Buffer = rapidjson::StringBuffer;
+	using Writer = rapidjson::Writer<Buffer>;
+	auto buffer = Buffer();
+	auto writer = Writer(buffer);
+	writer.StartObject();
+	writer.Key("keys");
+	writer.StartArray();
+	for (const auto& [keyName, key] : keys) {
+		if (!encodeKey(writer, keyName, key)) {
+			return {};
+		}
+	}
+	writer.EndArray();
+	writer.EndObject();
+	auto buf = new (arena) uint8_t[buffer.GetSize()];
+	::memcpy(buf, buffer.GetString(), buffer.GetSize());
+	return StringRef(buf, buffer.GetSize());
+}
+
+void forceLinkJsonWebKeySetTests() {}
+
+void testPublicKey(PrivateKey (*factory)()) {
```

Comment:
A rather simple name in a global namespace, consider moving into an anonymous namespace. The same for testPrivateKey

Reply:
Done!